### PR TITLE
fix(EN-2011): derive the signing cascade from effective ordered key state

### PR DESCRIPTION
**Pull request by Azorlogh**
## What changed

`RevokeSigningKey(cascade=true)` now removes a child that was revoked and then re-registered earlier in the same signed batch. `GetSigningKeyChildren` folds the batch's pending signing-key updates **in order** (last entry per key ID wins) and answers from that effective state, instead of suppressing any key that had a removal anywhere in the batch.

A key therefore belongs to the parent its latest registration named. The checker's cascade re-derivation drops the three per-proposal maps that existed only to mirror the old order-blind read.

## Why

In a batch signed by `P` containing, in order, `revoke(C)` / `register(C, new public key)` / `cascade-revoke(P)`, `C` survived the cascade and its replacement key still authenticated. That violates the cascade contract in [`docs/ops/signing.md`](docs/ops/signing.md), which promises cascade removes the entire descendant subtree.

`WriteSet.Merge` replays `pendingSigningKeyUpdates` in slice order, but `GetSigningKeyChildren` collected every removal in that slice into an unordered set and filtered both the committed children and the pending additions through it. The removal at index 0 suppressed `C` regardless of the registration at index 1, so the cascade staged no removal for it and `Merge` replayed `[del C, save C→P, del P]` with no fourth entry.

Revocation's only persistent representation is row absence, and signature verification looks up the requested key's own public key without walking to the parent — so deleting `P` did nothing to disable the surviving `C`.

Jira: [EN-2011](https://formance-team.atlassian.net/browse/EN-2011)

## Product / operational motivation

**Need:** an operator revoking a compromised key with `--cascade` must be able to rely on the whole subtree being gone.

**Current limitation:** a single atomic batch could leave a descendant alive with a working, attacker-chosen replacement key, and `cascaded_key_ids` recorded nothing — so neither the API response nor the audit trail showed the survivor. The same three operations submitted as separate batches already revoked it correctly, so the outcome depended on batch packing.

**Requirement / constraint:** after a cascade revoke, no key that is a descendant of the target at apply time authenticates, on the live path, after restart, and after cross-cluster restore. The batch boundary must not change which keys survive.

**Evidence:** [`docs/ops/signing.md`](docs/ops/signing.md) § Cascade Revocation; regression guards in [`internal/infra/state/signing_cascade_test.go`](internal/infra/state/signing_cascade_test.go).

**Durable repository evidence:** [`docs/technical/architecture/subsystems/admission/signing.md`](docs/technical/architecture/subsystems/admission/signing.md) § Cascade and the effective parent.

## Technical decision

**Decision:** derive the cascade's child lookup from an in-order fold of `pendingSigningKeyUpdates`, the same sequence `Merge` replays. Determinism without a sort: committed children arrive sorted from `keystore.GetChildren`, pending ones follow in slice order, which is fixed by the applied proposal.

**Why now / why proportionate:** the write path was already ordered; only the read was not. The fix is local to one function, adds no state to the FSM apply path, and makes the checker *simpler* — its expected-key map is already the effective ordered state, so `descendantsOf` can walk it directly.

**Alternatives considered:** a maintained overlay map on the `WriteSet`, rejected — it duplicates the slice as a second source of truth that `Reset` must keep in sync, and it cannot be iterated deterministically anyway, so the slice would still be needed for ordering.

## Risk

**MEDIUM**

Changes chain-hashed `cascaded_key_ids` for batches that re-register a key, and changes who survives a cascade in three reachable shapes (below). v3 is unreleased, so there is no compatibility surface, but this is signing authority.

## Validation

- [x] `bash scripts/agent-check`
- [x] Targeted tests: `internal/infra/state`, `internal/application/check`, `internal/infra/backup`, `internal/domain/processing`, `internal/application/admission`, `internal/domain/crypto`
- [x] Full suite / broader validation: `AI_REVIEW_BASE_SHA=2b52ff546 bash scripts/agent-check-pr` → selected and passed `agent-check-full` (re-run after rebasing onto `release/v3.0`)

The regression guards were verified against the pre-fix implementation: reverting only `GetSigningKeyChildren` fails `TestSigningCascadeReachesAReregisteredChild/one_batch`, `TestSigningCascadeSurvivesRestart` and three `FollowsTheEffectiveParent` cases, while `one_batch_per_order` passes both ways — confirming the batch boundary was the variable.

Coverage drives the real processor through `NewScopeFactory` → `WriteSet` → `Merge` with real Ed25519 keys, and probes authentication the way `resolveBatch` does (key-store lookup, then verify): the reported sequence, the split-batch control, restart, restore from a checkpoint plus non-empty delta, batch rollback, a rejected order, cycles, and the admission boundary where a revoked `P`'s original envelope is refused at the key lookup rather than at the idempotency gate.

## Architecture / behavior impact

**FSM apply:** `GetSigningKeyChildren` is read during Raft apply on every replica. The fold is a pure function of the proposal's own pending slice plus committed state — no node-local state, no new Pebble reads, no coverage-gate change.

**Behaviour changes inside a single batch**, all following from "the last registration decides":

| Earlier in the same batch | Before | After |
|---|---|---|
| `child` revoked, then re-registered under `parent` | survives | **revoked** (the bug) |
| `child` re-registered under a different parent | revoked | **survives** |
| `child` re-registered as a root | revoked | **survives** |
| an in-batch edge to `parent` superseded by a later registration | revoked | **survives** |

Rows 2–4 are not the reported defect but fall out of the same rule, and they make in-batch reassignment agree with the cross-batch case, which already behaved this way.

**Restore:** `RebuildDelta` is unchanged — it already replays logs in order and deletes `cascaded_key_ids` verbatim, so it inherits the corrected set.

**Not changed:** `RegisterSigningKey` keeps its upsert behaviour and batch-signer-derived parent, per the ticket.

## Review focus

1. **The fold's determinism** in `GetSigningKeyChildren` — the pending half relies on `pendingSigningKeyUpdates` order being identical on every replica, since the result reaches `cascaded_key_ids`.
2. **The checker collapse** in `internal/application/check/signing.go`: `proposalParents`, `proposalEdges`, `proposalRevoked`, `beginProposal` and `ensureProposalSnapshot` are removed. The claim is that they only ever reproduced the FSM's order-blind read, and that `v.keys` is already the effective ordered state. Worth an independent check that no projection verification is weakened.
3. **Rows 2–4 of the behaviour table** — agreed as in scope, but they are a deliberate widening beyond the reported sequence.

## Known concerns

[EN-2012](https://formance-team.atlassian.net/browse/EN-2012) (Backlog, unassigned) records a decision to reject `RegisterSigningKey` for an ID that is **still present**. When it lands, rows 2–4 above become unreachable through the API and the tests covering them should be deleted or rewritten in `revoke`-then-`register` form. This PR deliberately stays out of the registration path — the only change to `processor_signing.go` is a comment in the revoke path — and the EN-2011 sequence itself is unaffected, because `C` is absent at the middle operation. Coordinated with EN-2012's own note to that effect.

[EN-2011]: https://formance-team.atlassian.net/browse/EN-2011?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/ops/signing.md [hunk 1]
```diff
@@ -113,6 +113,27 @@ When `--cascade` is used, the revocation log includes a `cascaded_key_ids` field
...
+
+A cascade removes every key that is a descendant of the target **at the moment the revoke is applied**, using each key's current parent — the one its most recent registration assigned.
+
+Because registration is an upsert, a single signed batch can move keys around before the revoke in that same batch runs. The rule is that the last registration wins:
+
+| Earlier in the same batch | `revoke --cascade parent` |
+|---|---|
+| `child` revoked, then re-registered under `parent` | `child` **is** revoked — the re-registration put it back in the subtree |
+| `child` re-registered under a different parent | `child` **survives** — it left the subtree before the revoke ran |
+| `child` re-registered with no parent (becomes a root) | `child` **survives** |
```
Comment (NumaryBot) [inline docs/ops/signing.md]: <sub>OpenCodeReview · independent Codex pass · comment-only</sub>

🟠 **[major] Root reassignment is unavailable in a signed batch**

Admission always assigns the batch signer as the registration’s parent. Unsigned registration is rejected while keys exist, so an operator cannot perform the recommended root reassignment before revoking the parent. Remove this operational option or distinguish internal FSM behavior from API-accessible operations. The architecture document repeats this unsupported signed-batch claim.

File: docs/ops/signing.md [hunk 2]
```diff
@@ -113,6 +113,27 @@ When `--cascade` is used, the revocation log includes a `cascaded_key_ids` field
...
+#### Which keys a cascade reaches
+
+A cascade removes every key that is a descendant of the target **at the moment the revoke is applied**, using each key's current parent — the one its most recent registration assigned.
+
+Because registration is an upsert, a single signed batch can move keys around before the revoke in that same batch runs. The rule is that the last registration wins:
+
+| Earlier in the same batch | `revoke --cascade parent` |
+|---|---|
+| `child` revoked, then re-registered under `parent` | `child` **is** revoked — the re-registration put it back in the subtree |
+| `child` re-registered under a different parent | `child` **survives** — it left the subtree before the revoke ran |
```
Comment (NumaryBot) [inline docs/ops/signing.md]: <sub>OpenCodeReview · independent Codex pass · comment-only</sub>

🟠 **[major] A different parent does not necessarily escape the cascade**

The replacement parent must be outside the target’s entire subtree. For example, with `P → Q` and `P → C`, a batch signed by `Q` that re-registers `C` and cascade-revokes `P` still removes `C` through `Q`. The unconditional survival claim and accompanying operational advice can therefore cause unintended credential loss. Add the outside-subtree condition here and in the equivalent architecture and legacy descriptions.

**Review by NumaryBot — COMMENTED** [review]
OpenCodeReview's independent Codex pass added 2 comment-only findings.

These comments do not affect the active verdict or commit status.

File: docs/ops/signing.md (hunk 1 above)
Comment (Azorlogh) [reply docs/ops/signing.md]: Correct — `authorizeUnsignedBatch` admits an unsigned registration only while the key store is empty, so an empty parent is unreachable through the API once any key exists. Dropped the root option from the operational advice in both documents and stated that the new parent is always the batch signer; kept a note that the FSM still resolves an empty parent, since recovery and `RebuildDelta` rebuild roots from parent-less rows. c3b04b0bd

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: docs/ops/signing.md (hunk 2 above)
Comment (Azorlogh) [reply docs/ops/signing.md]: Confirmed with a test: `P → Q`, `P → C`, then re-register `C` under `Q` and cascade-revoke `P` gives `cascaded_key_ids=[Q C]` and `C` no longer authenticates. Added the outside-subtree condition to the table (now a row each way) in the ops, architecture and legacy documents, plus `TestSigningCascadeReachesThroughAReassignedParent` covering it. c3b04b0bd

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Review by NumaryBot — COMMENTED** [review]
NumaryBot review complete: no remaining inline findings.

Resolved 2 stale NumaryBot review threads (0 fixed, 2 outdated).

Summary: https://github.com/formancehq/ledger/pull/2074#issuecomment-5699963939

File: internal/application/check/signing.go [hunk 3]
```diff
@@ -227,43 +139,16 @@ func (v *signingVerifier) applyOrder(order *raftcmdpb.Order) {
...
-// while absence is a point-in-time fact, so revoke(X) + register(X under P) +
-// cascade-revoke(P) in ONE proposal has the FSM keep X — its removal excludes it
-// from the cascade even though the registration came after — while a walk driven
-// by v.keys alone follows the reinstated edge and reports a false mismatch against
-// the surviving row.
-//
-// The exclusion has to skip the candidate ENTIRELY rather than just drop it from
-// the result: GetSigningKeyChildren filters it out of what it returns, so the FSM's
-// BFS never recurses through it and its own subtree survives too.
+// The set of EDGES is what has to match, and v.keys already holds it. Both sides
```
Comment (NumaryBot) [inline internal/application/check/signing.go:145]: <sub>OpenCodeReview · independent Codex pass · comment-only</sub>

🟠 **[major] Update the checker’s documented cascade contract**

`docs/technical/architecture/subsystems/checker/checker.md:81-91` still requires the removed `proposalParents`, `proposalEdges`, and `proposalRevoked` bookkeeping and explicitly claims that revoke/re-register/cascade leaves the child alive. This now contradicts HEAD and presents the fixed behavior as a checker error. Update that authoritative section to describe the ordered effective-state fold and remove the obsolete requirements.

**Review by NumaryBot — COMMENTED** [review]
OpenCodeReview's independent Codex pass added 1 comment-only finding.

These comments do not affect the active verdict or commit status.

File: internal/infra/backup/rebuild_test.go [hunk 4]
```diff
@@ -1518,6 +1518,103 @@ func TestRebuildDelta_ReplaysRevokeSigningKey(t *testing.T) {
...
+			require.NoError(t, state.SaveSigningKey(batch, "P", parentPub, ""))
+			require.NoError(t, state.SaveSigningKey(batch, "C", childPub, "P"))
+
+			for _, log := range tt.delta {
+				require.NoError(t, batch.SetProto(coldLogKey(log.GetSequence()), log))
+			}
+
+			require.NoError(t, batch.Commit())
+
+			require.NoError(t, RebuildDelta(context.Background(), testLogger(), store, 0, 0))
```
Comment (NumaryBot) [inline internal/infra/backup/rebuild_test.go:1589]: 🟠 **[major] Prove cascade restore parity through the actual export path**

This test supplies handcrafted logs with the corrected `cascaded_key_ids` and calls `RebuildDelta` directly, so it cannot detect a failure to carry the live cascade result through checkpoint/export/restore or a disagreement with the checker. The [incremental restore contract:88-100](docs/technical/architecture/subsystems/backup/incremental-restore-contract.md#L88-L100) explicitly requires more than this isolated unit test: apply the affected batch after a checkpoint, restore a non-empty export through `ApplyExportsAndRebuild`, compare against live state, and run `CheckStore`. Add that cross-lifecycle regression while retaining this focused test.

**Review by NumaryBot — COMMENTED** [review]
NumaryBot posted 1 new inline finding.

Summary: https://github.com/formancehq/ledger/pull/2074#issuecomment-5699963939

File: internal/application/check/signing.go (hunk 3 above)
Comment (Azorlogh) [reply internal/application/check/signing.go:145]: Right, I missed that document. Rewrote the section to describe the ordered effective-state fold both sides now share, dropped the `proposalParents` / `proposalEdges` / `proposalRevoked` requirements, and corrected the reassignment case to note that a key only escapes the cascade when its new parent is outside the revoked subtree. The superseded model is kept as a short historical note so the position-blind read is not reintroduced. 289ddf941

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: internal/infra/backup/rebuild_test.go (hunk 4 above)
Comment (Azorlogh) [reply internal/infra/backup/rebuild_test.go:1589]: Agreed, the isolated test could only prove the replay half. Added `TestBackup_SigningCascadeRestoreParity`: a real FSM registers P, C and an unrelated root, `RunBackup` takes a full checkpoint while C is still live, the reported batch is applied after it, and the manifest's non-empty delta is restored through `ApplyExportsAndRebuild`. Restored signing rows are compared against the live source, and `Check` must report no findings. Reverting `GetSigningKeyChildren` to the position-blind read fails it at the live-apply assertion. Kept the focused `RebuildDelta` test. cf2792743

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Review by NumaryBot — COMMENTED** [review]
NumaryBot review complete: no remaining inline findings.

Resolved 1 stale NumaryBot review thread (1 fixed, 0 outdated).

Summary: https://github.com/formancehq/ledger/pull/2074#issuecomment-5699963939

File: internal/infra/backup/rebuild_test.go [hunk 5]
```diff
@@ -1518,6 +1518,103 @@ func TestRebuildDelta_ReplaysRevokeSigningKey(t *testing.T) {
...
+
+			for _, keyID := range tt.wantPresent {
+				require.Contains(t, keys, keyID, "key %q must survive the restore", keyID)
+			}
+
+			for _, keyID := range tt.wantAbsent {
+				require.NotContains(t, keys, keyID,
+					"revoked key %q must not be resurrected by the restore", keyID)
+			}
+
```
Comment (NumaryBot) [inline internal/infra/backup/rebuild_test.go:1608]: Guard asserts on `keys["C"]` but keys off `wantPresent`

The `if len(tt.wantPresent) &gt; 0` block asserts `keys["C"].PublicKey` and `ParentKeyID` regardless of whether `"C"` is in `wantPresent`. A future case expecting a different surviving key would fail on `C`'s zero-value row with a misleading message. Gate the assertions on `"C"` membership in `wantPresent` or move them into the per-case table.

File: internal/infra/backup/signing_restore_parity_test.go [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+	result, err := machine.ApplyEntries(context.Background(), store, &raftpb.Entry{
+		Index: new(index),
+		Term:  proto.Uint64(1),
+		Type:  new(raftpb.EntryNormal),
+		Data:  data,
+	})
+	require.NoError(t, err)
+
+	for _, r := range result.Results {
+		require.NoError(t, r.Error)
```
Comment (NumaryBot) [inline internal/infra/backup/signing_restore_parity_test.go:102]: Duplicated signing-order constructors across three test packages

`signingRegisterOrder`/`signingRevokeOrder` duplicate the same shape as `registerOrder`/`revokeOrder` in `internal/infra/state/signing_cascade_test.go` and `registerSigningKeyOrder`/`revokeSigningKeyOrder` in `internal/application/check/signing_test.go` (Duplicated Code, judgement call). A small shared test helper package would keep the three copies from drifting as the order shapes evolve.

**Review by NumaryBot — COMMENTED** [review]
Findings

File: internal/infra/backup/rebuild_test.go (hunk 5 above)
Comment (Azorlogh) [reply internal/infra/backup/rebuild_test.go:1608]: Right — gated on `"C"` being in `wantPresent`. 25474e51f

File: internal/infra/backup/signing_restore_parity_test.go (hunk 6 above)
Comment (Azorlogh) [reply internal/infra/backup/signing_restore_parity_test.go:102]: Leaving these as-is. Sharing them needs a non-test package under `internal/` so three test packages can import it, which puts test-only constructors in the production tree; there is no such precedent here (`tests/e2e/testutil` is e2e-only). Two 14-line literals per package is the cheaper trade.

File: internal/infra/state/write_set.go [hunk 7]
```diff
@@ -1266,39 +1266,71 @@ func (b *WriteSet) RemoveSigningKey(keyID string) {
...
+// effectiveSigningKeys folds pendingSigningKeyUpdates in slice order, the last
+// entry for a key ID winning, so a caller sees the state Merge will produce.
+//
+// Order is the whole point. The slice mixes additions and removals, and a key ID
+// may appear several times in one proposal — registration is an upsert, so
+// [remove C, save C under P] leaves C under P while the reverse leaves C gone.
+// Merge replays the slice top to bottom, so only a positional fold agrees with
+// what the proposal actually commits.
+func (b *WriteSet) effectiveSigningKeys() map[string]effectiveSigningKey {
+	effective := make(map[string]effectiveSigningKey, len(b.pendingSigningKeyUpdates))
```
Comment (NumaryBot) [inline internal/infra/state/write_set.go:1285]: **🔵 [suggestion] Cascade BFS re-folds all pending signing-key updates on every step**

`GetSigningKeyChildren` calls `effectiveSigningKeys()` on every invocation, and `processRevokeSigningKey` invokes it once per BFS node, so one cascade revoke costs O(descendants × pendingSigningKeyUpdates) map builds. Signing batches are tiny today and output is identical, so this is a performance judgement call, not a defect. Folding once per revoke (e.g. an optional precomputed effective map passed down) would make the cost linear. (Reported by both the spec and standards reviews.)

File: internal/infra/state/signing_cascade_test.go [hunk 8]
```diff
@@ -0,0 +1,572 @@
...
+				"the live key store and the persisted rows are written by one loop and must agree")
+
+			cascaded := payloads[len(payloads)-1].GetRevokeSigningKey().GetCascadedKeyIds()
+			require.Equal(t, []string{"C"}, cascaded,
+				"cascaded_key_ids is chain-hashed and replayed verbatim on restore, so it must name C")
+		})
+	}
+}
+
+// TestSigningCascadeSurvivesRestart pins that a replica which rebuilt its key
```
Comment (NumaryBot) [inline internal/infra/state/signing_cascade_test.go:267]: **🟡 [minor] TestSigningCascadeSurvivesRestart never rebuilds the key store**

The test asserts only the persisted rows and the un-restarted in-memory `machine.keyStore`: `authenticates` runs against the machine that applied the batch, never one rebuilt via the recovery path. Its name and comment claim invariant #1/#2 restart parity, but the claim that recovery rebuilds the key store from persisted rows and nothing else is asserted about, not exercised. Recreate a machine/key store from the same data directory and assert it agrees with the un-restarted one.

**Review by NumaryBot — COMMENTED** [review]
Findings

File: internal/infra/state/signing_cascade_test.go (hunk 8 above)
Comment (Azorlogh) [reply internal/infra/state/signing_cascade_test.go:267]: Good catch — it asserted the claim instead of exercising it. The test now boots a second Machine over the same store through `Recovery` and compares both the key set and the parent edges against the un-restarted one. 1bb870ec2

File: internal/infra/state/write_set.go (hunk 7 above)
Comment (Azorlogh) [reply internal/infra/state/write_set.go:1285]: Leaving as-is. Passing a precomputed map down means changing the `processing.Scope` interface (plus `skipSafeScope` and the generated mock) for a path where the fold is over a handful of entries and most proposals stage none. Caching it on the WriteSet would add a second source of truth next to the pending slice that `Reset` has to keep in sync — the invalidation bug that would eventually cause is worth more than the allocations.

File: internal/infra/state/write_set.go [hunk 9]
```diff
@@ -1266,39 +1266,71 @@ func (b *WriteSet) RemoveSigningKey(keyID string) {
...
+// GetSigningKeyChildren returns the key IDs whose parent is keyID in the state
+// the proposal has built so far: the committed children it has not touched, plus
+// every key whose folded parent is keyID.
+//
+// The result feeds the revocation cascade, and hence the chain-hashed
+// RevokedSigningKeyLog.cascaded_key_ids, so it must be identical on every replica
+// (invariant #2). Both halves are ordered without a sort: keystore.GetChildren
+// returns sorted committed children, and the pending half walks
+// pendingSigningKeyUpdates, whose order is fixed by the applied proposal.
+func (b *WriteSet) GetSigningKeyChildren(keyID string) []string {
```
Comment (NumaryBot) [inline internal/infra/state/write_set.go:1306]: **🔵 [suggestion] Cascade fold map rebuilt on every BFS step**

`GetSigningKeyChildren` rebuilds the `effectiveSigningKeys()` fold over the whole `pendingSigningKeyUpdates` slice on every call, and `processRevokeSigningKey` invokes it once per visited node, so a cascade of D nodes over U staged updates costs O(D·U) map builds inside the Raft apply path. The fold cannot change between orders: compute it once per revoke and pass it through, or memoize it on the `WriteSet` and invalidate it in `AddSigningKey`/`RemoveSigningKey`. (Consolidates the spec and standards findings of the same defect.)

**Review by NumaryBot — COMMENTED** [review]
Findings

File: internal/infra/state/write_set.go (hunk 9 above)
Comment (Azorlogh) [reply internal/infra/state/write_set.go:1306]: stale — same finding as https://github.com/formancehq/ledger/pull/2074#discussion_r4045361272, answered there.

**Comment by Azorlogh** [comment]
Re: the `descendantsOf` suggestion in https://github.com/formancehq/ledger/pull/2074#issuecomment-5728036583 — leaving it as-is.

A parent→children index has to be maintained alongside `v.keys` and invalidated on every register and revoke the fold applies. That is a second derived view of the same state, inside the pass whose entire value is being an independent re-derivation of the signing projection from the audit chain; a stale index there produces a wrong descendant set rather than a slow one, and it would be compared against the store as if it were authoritative.

The walk is also bounded in practice: `v.keys` holds the cluster-wide signing keys, which is a forest of tens of entries, and the walk runs once per audited cascade revoke.

Same reasoning as the `GetSigningKeyChildren` fold in https://github.com/formancehq/ledger/pull/2074#discussion_r4045361272.

**Review by paul-nicolas — APPROVED** [review]
